mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,530 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2015 Cisco Systems, Inc. and others.  All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains a text Processing Pipeline."""
+import logging
+from typing import List, Dict
+import re
+import unicodedata
+
+from .preprocessors import Preprocessor, PreprocessorFactory, NoOpPreprocessor
+from .normalizers import Normalizer, NoOpNormalizer, NormalizerFactory
+from .tokenizers import Tokenizer, TokenizerFactory
+from .stemmers import Stemmer, StemmerFactory, NoOpStemmer
+
+from ..components._config import (
+    get_text_preparation_config,
+    DEFAULT_NORMALIZERS,
+    DEFAULT_EN_TEXT_PREPARATION_CONFIG,
+    get_language_config,
+    ENGLISH_LANGUAGE_CODE
+)
+from ..constants import UNICODE_SPACE_CATEGORY
+
+logger = logging.getLogger(__name__)
+
+
+# Regex Pattern to capture MindMeld entities ("{entity_text|entity_type|optional_role}")
+MINDMELD_ANNOTATION_PATTERN = re.compile(r"\{([^\}\|]*)\|[^\{]*\}")
+
+
+class TextPreparationPipelineError(Exception):
+    pass
+
+
+class TextPreparationPipeline:
+    """Pipeline Class for MindMeld's text processing."""
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        stemmer: Stemmer = None,
+        preprocessors: List[Preprocessor] = None,
+        normalizers: List[Normalizer] = None,
+        language: str = ENGLISH_LANGUAGE_CODE,
+    ):
+        """Creates a Pipeline instance."""
+        self.language = language
+        self.preprocessors = preprocessors or [NoOpPreprocessor()]
+        self.normalizers = normalizers or [NoOpNormalizer()]
+        self.tokenizer = tokenizer
+        self.stemmer = stemmer or NoOpStemmer()
+
+        if self.tokenizer is None:
+            raise TextPreparationPipelineError("Tokenizer cannot be None.")
+
+    def preprocess(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            normalized_text (str): Normalized Text.
+            forward_map (Dict): Mapping from raw text to modified text.
+            backward_map (Dict): Reverse mapping from modified text to raw text.
+        """
+        preprocessed_text = text
+        for preprocessor in self.preprocessors:
+            preprocessed_text = TextPreparationPipeline.modify_around_annotations(
+                text=preprocessed_text, function=preprocessor.process
+            )
+        return preprocessed_text
+
+    def normalize(self, text):
```

Comment:
I think this function should be named `tokenize_and_normalize`

Reply:
Here is how I would separate out the functions:
1. Tokenize: str -> list[str]
2. Normalize: str -> str
3. TokenizeAndNormalize: str -> list[str]
The callers of these primitive functions can then do things like `` `.join`

Reply:
As discussed:

Tokenize: str -> list[str]
Normalize: str -> str

TokenizeAndNormalize: str -> list[Dict]